File: server/src/web/jwtSecret.ts
import { existsSync } from 'node:fs'
import { chmod, mkdir, readFile, writeFile } from 'node:fs/promises'
import { randomBytes } from 'node:crypto'
import { dirname, join } from 'node:path'
import { z } from 'zod'
import { configuration } from '../configuration'

const jwtSecretFileSchema = z.object({
    secretBase64: z.string()
})

export async function getOrCreateJwtSecret(): Promise<Uint8Array> {
    const secretFile = join(configuration.dataDir, 'jwt-secret.json')

    if (existsSync(secretFile)) {
        await chmod(secretFile, 0o600).catch(() => {})
        const raw = await readFile(secretFile, 'utf8')
        const parsed = jwtSecretFileSchema.parse(JSON.parse(raw))
        const bytes = new Uint8Array(Buffer.from(parsed.secretBase64, 'base64'))
        if (bytes.length !== 32) {
            throw new Error(`Invalid JWT secret length in ${secretFile}`)
        }
        return bytes
    }

    const secretBytes = new Uint8Array(randomBytes(32))
    const dir = dirname(secretFile)
    if (!existsSync(dir)) {
        await mkdir(dir, { recursive: true, mode: 0o700 })
    }

    const payload = {
        secretBase64: Buffer.from(secretBytes).toString('base64')
    }
    await writeFile(secretFile, JSON.stringify(payload, null, 4), { mode: 0o600 })
    await chmod(secretFile, 0o600).catch(() => {})

    return secretBytes
}

File: hub_go/internal/config/jwt_secret.go
package config

import (
    "crypto/rand"
    "encoding/base64"
    "encoding/json"
    "errors"
    "os"
    "path/filepath"
)

type jwtSecretFile struct {
    SecretBase64 string `json:"secretBase64"`
}

func LoadOrCreateJWTSecret(dataDir string) ([]byte, error) {
    secretFile := filepath.Join(dataDir, "jwt-secret.json")

    raw, err := os.ReadFile(secretFile)
    if err == nil {
        var parsed jwtSecretFile
        if err := json.Unmarshal(raw, &parsed); err != nil {
            return nil, err
        }
        decoded, err := base64.StdEncoding.DecodeString(parsed.SecretBase64)
        if err != nil {
            return nil, err
        }
        if len(decoded) != 32 {
            return nil, errors.New("invalid JWT secret length")
        }
        return decoded, nil
    }

    if !errors.Is(err, os.ErrNotExist) {
        return nil, err
    }

    secret := make([]byte, 32)
    if _, err := rand.Read(secret); err != nil {
        return nil, err
    }

    payload := jwtSecretFile{SecretBase64: base64.StdEncoding.EncodeToString(secret)}
    encoded, err := json.MarshalIndent(payload, "", "    ")
    if err != nil {
        return nil, err
    }

    if err := os.MkdirAll(filepath.Dir(secretFile), 0o700); err != nil {
        return nil, err
    }

    if err := os.WriteFile(secretFile, encoded, 0o600); err != nil {
        return nil, err
    }

    return secret, nil
}
